lib/cosmos/tools/launcher/launcher_config.rb

```diff
@@ -42,132 +44,162 @@ def initialize(filename)
-              variable_params = nil
-              if params.length > 3
-                raise parser.error("Unbalanced variable params for #{params[0]}") if (params.length % 2) != 1
-                variable_params = []
-                params[3..-1].each_slice(2) { |variable_parameter| variable_params << variable_parameter }
-              end
-              @items << [:TOOL, params[0], format_shell_command(params[1]), true, ConfigParser.handle_nil(params[2]), variable_params]
-            end
+        parse_file(filename)
+      else
+        raise "Launcher configuration file does not exist: #{filename}"
+      end
+    end # def initialize
 
-          when 'MULTITOOL_START'
-            parser.verify_num_parameters(1, 2, "MULTITOOL_START <Button Text> <Icon Filename (optional)>")
-            multitool = true
-            multitool_text = params[0]
-            multitool_icon_filename = ConfigParser.handle_nil(params[1])
-            multitool_icon_filename = 'multi.png' unless multitool_icon_filename
-            multitool_settings = []
-
-          when 'MULTITOOL_END'
-            parser.verify_num_parameters(0, 0, "MULTITOOL_END")
-            @items << [:MULTITOOL, multitool_text, multitool_settings, true, multitool_icon_filename, nil]
-            multitool = false
-            multitool_text = nil
-            multitool_icon_filename = nil
-            multitool_settings = nil
-
-          when 'DELAY'
-            if multitool
-              parser.verify_num_parameters(1, 1, "DELAY <Delay in seconds>")
-              multitool_settings << [:DELAY, Float(params[0]), true]
-            else
-              raise parser.error("DELAY keyword only valid within MULTITOOL")
+    # Create a ConfigParser and parse all the lines in the configuration file
+    #
+    # @param filename [String] Name of the configuration file to parse
+    def parse_file(filename)
+      multitool = nil
+
+      # Loop over each line of the configuration file
+      parser = ConfigParser.new
+      parser.parse_file(filename) do |keyword, params|
+        # Handle each keyword
+        case keyword
+
+        when 'TOOL'
+          parse_tool(parser, params, multitool)
+
+        when 'MULTITOOL_START'
+          parser.verify_num_parameters(1, 2, "MULTITOOL_START <Button Text> <Icon Filename (optional)>")
+          multitool = OpenStruct.new
+          multitool.text = params[0]
+          multitool.icon_filename = ConfigParser.handle_nil(params[1])
+          multitool.icon_filename = 'multi.png' unless multitool.icon_filename
+          multitool.settings = []
+
+        when 'MULTITOOL_END'
+          parser.verify_num_parameters(0, 0, "MULTITOOL_END")
+          raise parser.error("No TOOLs defined within the MULTITOOL") if multitool.settings.select {|setting| setting[0] == :TOOL }.empty?
+          @items << [:MULTITOOL, multitool.text, multitool.settings, true, multitool.icon_filename, nil]
```

Comment:
I think a future refactor might make the @items array an object but I didn't want to tackle that.